shorthand serialization for background and mask should not include "initial" in the middle of the value

darinadler opened the pull request:
#### a0488d1bf07c46a94b77ef9a721b5de4e20a92c1
<pre>
shorthand serialization for background and mask should not include &quot;initial&quot; in the middle of the value
<a href="https://bugs.webkit.org/show_bug.cgi?id=242775">https://bugs.webkit.org/show_bug.cgi?id=242775</a>
rdar://97405629

Reviewed by Tim Nguyen.

Moved the code that handles CSS-wide keywords into the commonShorthandChecks function, and got rid
of many of the attempts to handle it in various shorthand-specific functions. To do this without
causing regressions, had to make the handling of the implicit flag more consistent; we&apos;ll be
removing the implicit flag as soon as we can, but we can&apos;t remove it until we have serialization
that understands initial values of the longhands for all the shorthands.

* LayoutTests/css3/flexbox/css-properties-expected.txt: Updated since setting a longhand to
&quot;initial&quot; means the shorthand will serialize as the empty string.
* LayoutTests/css3/flexbox/css-properties.html: Ditto.

* LayoutTests/fast/css/cssText-shorthand-expected.txt: Updated since an incorrect test is now
failing. I think this test is a duplicate of the test below,
imported/w3c/web-platform-tests/css/cssom/shorthand-values-expected.txt, so we may want to
delete this one.

* LayoutTests/fast/css/remove-shorthand-expected.txt: Updated for a progression. Since
-webkit-mask is a legacy shorthand, it should serialize as the empty string, and so it should not
have shown up in these test results.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks-expected.txt:
Updated for progressions. Some tests that were failing are now passig. Others are still failing,
but are closer to passing.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/flex-serialization-expected.txt:
One more test passing.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/overflow-serialization-expected.txt:
One more test passing.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/shorthand-serialization-expected.txt:
One more test passing.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/shorthand-values-expected.txt: Updated
since an incorrect test is now failing. The test expectation is wrong, and we should fix the test.
For now, I did not include the test fix here because the fixed test will still fail for a
different reason. Can handle that change separately.

* LayoutTests/platform/mac/editing/execCommand/find-after-replace-expected.txt: Updated for a
progression. This was improved by another recent change, not any of the changes here.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::commonShorthandChecks const): Added checks for CSS-wide keywords,
implementing two related rules: 1) If all longhands have the same CSS-wide keyword value,
serialize the shorthand as that value. 2) If some but not all longhands have a CSS-wide keyword
value, serialize as empty string. At this time there is still an exception to those rules for
initial implicit value.
(WebCore::StyleProperties::fontValue const): Removed CSS-wide keyword handling. Still need
similar code to handle system font shorthand.
(WebCore::StyleProperties::fontVariantValue const): Ditto.
(WebCore::StyleProperties::fontSynthesisValue const): Ditto.
(WebCore::StyleProperties::get2Values const): Removed special handling for inherit. Should
also be able to remove special handling of initial, will do that in a follow-up patch.
(WebCore::StyleProperties::getLayeredShorthandValue const): Removed CSS-wide keyword handling.
(WebCore::StyleProperties::getShorthandValue const): Removed CSS-wide keyword handling.
(WebCore::StyleProperties::getCommonValue const): Skip implicit initial values. Also removed a
comment that does not make sense referring to a CSSValue derived class we don&apos;t have.
(WebCore::StyleProperties::getAlignmentShorthandValue const): Streamlined a tiny bit.
(WebCore::StyleProperties::borderImagePropertyValue const): Removed CSS-wide keyword handling.
(WebCore::StyleProperties::borderRadiusShorthandValue const): Ditto.
(WebCore::StyleProperties::borderPropertyValue const): Ditto. Also skip implicit initial values.
(WebCore::StyleProperties::breakInsideShorthandValue const): Removed CSS-wide keyword handling.
(WebCore::StyleProperties::pageBreakValue const): Ditto.
(WebCore::StyleProperties::webkitColumnBreakValue const): Ditto.
(WebCore::StyleProperties::asTextInternal const): Skip implicit initial values unless they
are inherited. Removed the old code trying to do this in a less precise fashion.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::addProperty): Set the implicit flag for all implicit initial values.
Some call sites were passing in false for implicit, but that was clearly unintentional. It was
much easier to make this change here than the plumb through the implicit flag correctly.

Canonical link: <a href="https://commits.webkit.org/257429@main">https://commits.webkit.org/257429@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/968f3c1652d737894a54f27c6421d3dc4f61dfd6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98842 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8053 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32037 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108258 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168515 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102796 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8603 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85420 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91375 "Built successfully") | [❌ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106246 "Hash 968f3c16 for PR 7025 does not build (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104525 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/6518 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90118 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33542 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88340 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21462 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/76407 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1965 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22989 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1874 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/45416 "Found 1 new test failure: media/video-inactive-playback.html (failure)") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5112 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6830 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42452 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3272 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/cssom/shorthand-values-expected.txt:
```diff
@@ -14,7 +14,7 @@ PASS The serialization of overflow-x: scroll; overflow-y: hidden; should be cano
 PASS The serialization of outline-width: 2px; outline-style: dotted; outline-color: blue; should be canonical.
 PASS The serialization of margin-top: 1px; margin-right: 2px; margin-bottom: 3px; margin-left: 4px; should be canonical.
-PASS The serialization of list-style-type: circle; list-style-position: inside; list-style-image: initial; should be canonical.
+FAIL The serialization of list-style-type: circle; list-style-position: inside; list-style-image: initial; should be canonical. assert_equals: expected "list-style: inside circle;" but got "list-style-type: circle; list-style-position: inside; list-style-image: initial;"
```
Can we update the WPT and upstream the change?

darinadler replied:
I plan to do that in another change later; I have it stashed elsewhere. The reason I didn’t do it now is that we won’t pass the revised test because we don’t yet support initial values properly.

darinadler replied:
Another option is that I revise the test so that it still tries to test with `initial` and just has an appropriate expectation.

nt1m replied:
Either is fine by me, I just don't want this to slip through the cracks.

darinadler replied:
I will follow through; this will not slip through the cracks.

Loirooriol commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -177,12 +193,14 @@ String StyleProperties::commonShorthandChecks(const StylePropertyShorthand& shor
             numSetFromShorthand += 1;
     }
+    if (specialKeyword)
+        return allSpecialKeywords ? nameString(*specialKeyword) : emptyString();
```
I think this could just be

```cpp
    if (specialKeyword)
        return nameString(*specialKeyword);
```

If you keep the invariant `allSpecialKeywords || !specialKeyword` by returning early like

```cpp
        if (propertyIndex == -1) {
            if (specialKeyword)
                return emptyString();
            allSpecialKeywords = false;
            continue;
        }
```
```cpp
        if (!value->isCSSWideKeyword() || value->isImplicitInitialValue()) {
            if (specialKeyword)
                return emptyString();
            allSpecialKeywords = false;
        } else {
            auto keyword = valueID(value);
            if (specialKeyword.value_or(keyword) != keyword || !allSpecialKeywords)
                return emptyString();
            specialKeyword = keyword;
            continue;
        }
```

And then avoid unnecessary iterations.

darinadler replied:
I did something similar.

Loirooriol commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -666,9 +664,6 @@ String StyleProperties::get2Values(const StylePropertyShorthand& shorthand) cons
-    if (start.isInherited() && end.isInherited())
-        return nameString(CSSValueInherit);
-
     if (start.value()->isInitialValue() || end.value()->isInitialValue()) {
         if (start.value()->isInitialValue() && end.value()->isInitialValue() && !start.isImplicit())
             return nameString(CSSValueInitial);
```
This can be removed too? Also for `StyleProperties::get4Values`, etc.

darinadler replied:
Yes, I will return and remove it later, but I prefer not to complicate this change further. Let me know if you object.

Loirooriol replied:
It's fine to do in a follow-up

Loirooriol commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -1718,6 +1642,9 @@ StringBuilder StyleProperties::asTextInternal() const
         Vector<CSSPropertyID> shorthands;
 
+        if (property.isImplicit())
+            continue;
```
This seems bad, I'm imagining

```
var {style} = document.createElement("div");
style.borderImage = "url()";
style.removeProperty("border-image-source");
```

Then `border-image-{slice,width,outset,repeat}` will be set to an implicit value and they will not appear in `cssText`?

Maybe check `property.value()->isImplicitInitialValue() && !CSSProperty::isInheritedProperty(propertyID)` instead? And then remove this code below, should help with https://bugs.webkit.org/show_bug.cgi?id=185953

```cpp
if (propertyID != CSSPropertyCustom && value == "initial"_s && !CSSProperty::isInheritedProperty(propertyID))
    continue;
```

darinadler replied:
Good point, I will look into this.

darinadler replied:
Why do I need to check `isInheritedProperty` at all?

darinadler replied:
I would like to remove that `CSSPropertyCustom` code, but maybe I or someone else can do it as part of the follow up cleanup and don’t have to do it at the same time I land this change?

Loirooriol replied:
> Why do I need to check `isInheritedProperty` at all?

So, I don't think initial values should be skipped at all. But there is this code for some reason (it's a workaround for some other bug if I recall correctly), and it's checking `!CSSProperty::isInheritedProperty`, so I would just keep that condition. Just tightening `value == "initial"_s` into `value()->isImplicitInitialValue()`.

But I think it's fine to not change `asTextInternal()` here and leave the change for a follow-up.

Loirooriol commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -1253,43 +1194,35 @@ String StyleProperties::borderRadiusShorthandValue(const StylePropertyShorthand&
-    if (isCSSWideKeyword(*valueID))
-        return nameString(*valueID);
-    switch (*valueID) {
+    auto keyword = propertyAsValueID(longhand).value_or(CSSValueInvalid);
```
Why does `propertyAsValueID` return a `std::optional<CSSValueID>` instead of falling back to `CSSValueInvalid`? It uses `valueID()`, which is already falling back to `CSSValueInvalid` for non-primitive values, so it doesn't make much sense to me.

darinadler replied:
The reason is that `std::nullopt` means "no property" and `CSSValueInvalid` means "property but its value is not an ID" and at least one caller takes advantage of that difference.

I agree that we could come up with a better design, but we can’t simply start returning `CSSValueInvalid` for missing properties without addressing this.

Loirooriol commented on LayoutTests/platform/mac/editing/execCommand/find-after-replace-expected.txt:
```diff
@@ -29,8 +29,6 @@ layer at (0,0) size 800x600
-                  text run at (0,0) width 26: "A B"
-                  text run at (25,0) width 5: " "
-                  text run at (29,0) width 26: "A B"
+                  text run at (0,0) width 55: "A B A B"
```
Why does this change happen?

darinadler replied:
The progression was caused by some other recent code change, not in this pull request. We can land this separately.

Loirooriol commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -1711,6 +1641,9 @@ StringBuilder StyleProperties::asTextInternal() const
         ASSERT(isLonghand(propertyID) || propertyID == CSSPropertyCustom);
         Vector<CSSPropertyID> shorthands;
 
+        if (property.value()->isImplicitInitialValue())
```
If you add this, then I think you may remove

```cpp
if (propertyID != CSSPropertyCustom && value == "initial"_s && !CSSProperty::isInheritedProperty(propertyID))
    continue;
```

Regardless, not convinced of spreading this bug into inherited properties:

```js
var {style} = document.createElement("div");
style.cssText = "text-emphasis: dot";
style.removeProperty("text-emphasis-style");
style.cssText;
// Expected:   "text-emphasis-color: currentcolor;"
// Current:    "text-emphasis-color: initial;"
// With patch: ""
```

Skipping it means that it won't survive `element.style.cssText += ""`, and since it's inherited it will change from `currentcolor` to the parent value, that may be different.

So this should check:

```cpp
if (property.value()->isImplicitInitialValue() && !CSSProperty::isInheritedProperty(propertyID))
```

darinadler replied:
When I tried `!CSSProperty::isInheritedProperty` I ran into tests that failed.

Loirooriol replied:
Can you provide an example?

darinadler replied:
I will try to figure out what to do next, given your worry about spreading this rule further. But in the end I think we need to move away from implicit initial fairly quickly and soon

Loirooriol replied:
I'm actually not sure why you need to modify `asTextInternal()`. `commonShorthandChecks()` still allows mixing implicit `initial` with other things, so the shorthand should still serialize.

Is it something like having some longhands set to an implicit `initial` and the other longhands set to an explicit `initial`? What about letting the shorthand serialize as `"initial"`?

darinadler replied:
I did what you suggested. It works great.

Loirooriol reviewed (approved):
Great! Informal LGTM

nt1m commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -178,6 +178,9 @@ void CSSPropertyParser::addProperty(CSSPropertyID property, CSSPropertyID curren
     // execCommand('FontSizeDelta')).
     ASSERT(isExposed(property, &m_context.propertySettings) || setFromShorthand || isInternal(property));
 
+    if (!implicit && value->isImplicitInitialValue())
+        implicit = true;
```
It would be cool to stop using this flag at some point.

darinadler replied:
We will!

darinadler committed:
shorthand serialization for background and mask should not include "initial" in the middle of the value
https://bugs.webkit.org/show_bug.cgi?id=242775
rdar://97405629

Reviewed by Tim Nguyen.

Moved the code that handles CSS-wide keywords into the commonShorthandChecks function, and got rid
of many of the attempts to handle it in various shorthand-specific functions. To do this without
causing regressions, had to make the handling of the implicit flag more consistent; we'll be
removing the implicit flag as soon as we can, but we can't remove it until we have serialization
that understands initial values of the longhands for all the shorthands.

* LayoutTests/css3/flexbox/css-properties-expected.txt: Updated since setting a longhand to
"initial" means the shorthand will serialize as the empty string.
* LayoutTests/css3/flexbox/css-properties.html: Ditto.

* LayoutTests/fast/css/cssText-shorthand-expected.txt: Updated since an incorrect test is now
failing. I think this test is a duplicate of the test below,
imported/w3c/web-platform-tests/css/cssom/shorthand-values-expected.txt, so we may want to
delete this one.

* LayoutTests/fast/css/remove-shorthand-expected.txt: Updated for a progression. Since
-webkit-mask is a legacy shorthand, it should serialize as the empty string, and so it should not
have shown up in these test results.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks-expected.txt:
Updated for progressions. Some tests that were failing are now passig. Others are still failing,
but are closer to passing.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/flex-serialization-expected.txt:
One more test passing.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/overflow-serialization-expected.txt:
One more test passing.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/shorthand-serialization-expected.txt:
One more test passing.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/shorthand-values-expected.txt: Updated
since an incorrect test is now failing. The test expectation is wrong, and we should fix the test.
For now, I did not include the test fix here because the fixed test will still fail for a
different reason. Can handle that change separately.

* LayoutTests/platform/mac/editing/execCommand/find-after-replace-expected.txt: Updated for a
progression. This was improved by another recent change, not any of the changes here.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::commonShorthandChecks const): Added checks for CSS-wide keywords,
implementing two related rules: 1) If all longhands have the same CSS-wide keyword value,
serialize the shorthand as that value. 2) If some but not all longhands have a CSS-wide keyword
value, serialize as empty string. At this time there is still an exception to those rules for
initial implicit value.
(WebCore::StyleProperties::fontValue const): Removed CSS-wide keyword handling. Still need
similar code to handle system font shorthand.
(WebCore::StyleProperties::fontVariantValue const): Ditto.
(WebCore::StyleProperties::fontSynthesisValue const): Ditto.
(WebCore::StyleProperties::get2Values const): Removed special handling for inherit. Should
also be able to remove special handling of initial, will do that in a follow-up patch.
(WebCore::StyleProperties::getLayeredShorthandValue const): Removed CSS-wide keyword handling.
(WebCore::StyleProperties::getShorthandValue const): Removed CSS-wide keyword handling.
(WebCore::StyleProperties::getCommonValue const): Skip implicit initial values. Also removed a
comment that does not make sense referring to a CSSValue derived class we don't have.
(WebCore::StyleProperties::getAlignmentShorthandValue const): Streamlined a tiny bit.
(WebCore::StyleProperties::borderImagePropertyValue const): Removed CSS-wide keyword handling.
(WebCore::StyleProperties::borderRadiusShorthandValue const): Ditto.
(WebCore::StyleProperties::borderPropertyValue const): Ditto. Also skip implicit initial values.
(WebCore::StyleProperties::breakInsideShorthandValue const): Removed CSS-wide keyword handling.
(WebCore::StyleProperties::pageBreakValue const): Ditto.
(WebCore::StyleProperties::webkitColumnBreakValue const): Ditto.
(WebCore::StyleProperties::asTextInternal const): Skip implicit initial values unless they
are inherited. Removed the old code trying to do this in a less precise fashion.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::addProperty): Set the implicit flag for all implicit initial values.
Some call sites were passing in false for implicit, but that was clearly unintentional. It was
much easier to make this change here than the plumb through the implicit flag correctly.

Canonical link: https://commits.webkit.org/257429@main
